src/sqlfluff/dialects/dialect_snowflake.py

```diff
@@ -2342,6 +2342,65 @@ class CreateStatementSegment(BaseSegment):
         ),
         Ref("IfNotExistsGrammar", optional=True),
         Ref("ObjectReferenceSegment"),
+        # Next set are Storage Integration statements
+        # https://docs.snowflake.com/en/sql-reference/sql/create-storage-integration.html
+        Sequence(
+            AnySetOf(
+                Sequence("TYPE", Ref("EqualsSegment"), "EXTERNAL_STAGE"),
+                Sequence("ENABLED", Ref("EqualsSegment"), Ref("BooleanLiteralGrammar")),
+                OneOf(
+                    Ref("S3StorageIntegrationParameters"),
+                    Ref("GCSStorageIntegrationParameters"),
+                    Ref("AzureStorageIntegrationParameters"),
+                ),
+                # Note Could result in mismatched IntegrationParams and AllowedLocations
+                # E.G. S3 IntegrationParameters would parse with GCS compliant paths
+                # Alternative would be to add these to IntegrationParameter Segments but
+                # I feel like that is harder to understand the match_grammar
+                # when comparing against Snowflake docs.
```

Comment:
It's not necessary to write 100% compliant parsing code. While it would be nice to only parse exactly what Snowflake would accept., you need to weigh up how likely the above is, whether it would be expected for a SQL linter to catch those, versus the complexity of the code you're adding to SQLFluff. We've lots of places where we are overly permissive as to what we allow.

Also how likely is it someone will add S3 in one place and GCS on another? And does Snowflake itself throw a parsing error in this case? Or a runtime error?

Saying that I don't think adding to the IntegrationParameters is that bad. A comment would explain it.

Or alternatively do something like this:

```py
                OneOf(
                    Sequence(
                        Ref("S3StorageIntegrationParameters"),
                        Ref("S3StorageAllowdLocations"),
                    ),
                    Sequence(
                        Ref("GCSStorageIntegrationParameters"),
                        Ref("GCSStorageAllowdLocations"),
                    ),
                    Sequence(
                        Ref("AzureStorageIntegrationParameters"),
                        Ref("AzureStorageAllowdLocations"),
                    ),
                ),
```

Reply:
In this case - it would only be a runtime error, but the parsing would be fine.

Given the above, do you think I should leave it as it is (with my comments removed) or I could also change to as you suggest (with slight edit):

```
 OneOf(
                    AnySetOf(
                        Ref("S3StorageIntegrationParameters"),
                        Ref("S3StorageAllowdLocations"),
                    ),
```

I think I'm happy to leave the code as it is - like you mention, we can parse it and so can Snowflake. It's only the equivalent to someone adding a bad S3 path that doesn't actually exist!

Let me know and I can update tomorrow :) Saw your other comments too, thanks super helpful - will implement them tomorrow morn!

Reply:
Leaving as is SGTM!